app/main/views/audits.py

```diff
@@ -83,23 +90,43 @@ def list_audits():
     if object_type:
         if object_type not in AUDIT_OBJECT_TYPES:
             abort(400, 'invalid object-type supplied')
-        if not object_id:
-            abort(400, 'object-type cannot be provided without object-id')
+
         model = AUDIT_OBJECT_TYPES[object_type]
         id_field = AUDIT_OBJECT_ID_FIELDS[object_type]
 
-        ref_object = model.query.filter(
-            id_field == object_id
-        ).first()
+        audits = audits.filter(AuditEvent.object.is_type(model))
 
-        if ref_object is None:
-            abort(404, "Object with given object-type and object-id doesn't exist")
+        if object_id:
+            ref_object = model.query.filter(
+                id_field == object_id
+            ).first()
 
-        audits = audits.filter(AuditEvent.object == ref_object)
+            if ref_object is None:
+                abort(404, "Object with given object-type and object-id doesn't exist")
 
+            audits = audits.filter(
+                AuditEvent.object_id == class_mapper(model).identity_key_from_instance(ref_object)[1][0]
+            )
     elif object_id:
         abort(400, 'object-id cannot be provided without object-type')
 
+    if earliest_for_each_object:
+        audits_subquery = audits.order_by(
+            AuditEvent.object_type,
+            AuditEvent.object_id,
+            AuditEvent.created_at,
+            AuditEvent.id,
+        ).distinct(
+            AuditEvent.object_type,
+            AuditEvent.object_id,
+        ).subquery()
+
+        audits = AuditEvent.query.join(audits_subquery, audits_subquery.c.id == AuditEvent.id)
+
+    latest_first = convert_to_boolean(request.args.get('latest_first'))
+    pseudo_asc, pseudo_desc, pseudo_nullslast, pseudo_nullslast = reversible_order_by_terms(latest_first)
```

Comment:
* No, it's my damn autocomplete being overzealous
* This is all so gone you won't remember it next week.